### ISSUE TITLE
[HOLD for payment 2024-10-24] [$250] Error when copilot adds a new contact method

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#50063]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.42-3
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Issue was found when executing this PR:** https://github.com/Expensify/App/pull/49150
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Log in the App
2. Add a copilot to the user A
3. Login as copilot
4. Go to Account settings.
5. Click on the account switcher
6. Switch to user A account
7. Click Contact method.
8. Click New contact method.
9. Enter a new email address
10. Click Next

## Expected Result:
The new email address is added as a new contact method.

## Actual Result:
Copilot can not add new contact method. Error message "An unexpected error occurred when trying to add the new login.". RD in LHN-Profile, Profile window-Contact method, and LHP.

## Workaround:
Unknown

## Platforms:

- [ ] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/9fbc47d6-034b-4dda-bd55-36af84193a60

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021842646552595795235</li>
        <li>Upwork Job ID: 1842646552595795235</li>
        <li>Last Price Increase: 2024-10-05</li>
<li>Automatic offers: </li>
<ul>
<li>c3024 | Reviewer | 104335384</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @anmurali</details>

**Comment by IuliiaHerets** [comment]
We think that this bug might be related to #wave-collect - Release 1

**Comment by IuliiaHerets** [comment]
@anmurali  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by ChavdaSachin** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Error when copilot adds a new contact method

### What is the root cause of that problem?
- Copilot does not have enough authorities to add new contact methods.

### What changes do you think we should make in order to solve the problem?
Option 1.
- Disable `New Contact Method` button if current user is a copilot.
```js
    const [account] = useOnyx(ONYXKEYS.ACCOUNT);
    const isActingAsDelegate = !!account?.delegatedAccess?.delegate ?? false;
    .
    .
    .
    <Button
        isDisabled = {isActingAsDelegate}
        large
        success
        text={translate('contacts.newContactMethod')}
        onPress={onNewContactMethodButtonPress}
        pressOnEnter
    />
```
https://github.com/Expensify/App/blob/3946b226d6367a719af5bc894546b95dc9b59fb4/src/pages/settings/Profile/Contacts/ContactMethodsPage.tsx#L125-L131
> optionally we could add a short message above `New Contact Method` button to inform user that - copilot could not add new contact method.

Option 2.
- Hide `New Contact Method` button if current user is copilot.
```js
    const [account] = useOnyx(ONYXKEYS.ACCOUNT);
    const isActingAsDelegate = !!account?.delegatedAccess?.delegate ?? false;
    .
    .
    .
    { !isActingAsDelegate && <Button 
        large
        success
        text={translate('contacts.newContactMethod')}
        onPress={onNewContactMethodButtonPress}
        pressOnEnter
    />}
```
https://github.com/Expensify/App/blob/3946b226d6367a719af5bc894546b95dc9b59fb4/src/pages/settings/Profile/Contacts/ContactMethodsPage.tsx#L125-L131


- In addition to both of the above options - wrap [NewContactMethodPage](https://github.com/Expensify/App/blob/3946b226d6367a719af5bc894546b95dc9b59fb4/src/pages/settings/Profile/Contacts/NewContactMethodPage.tsx) with `AccessOrNotFoundWrapper` to avoid copilot using deep link to access the page.
```js
    <AccessOrNotFoundWrapper
        shouldBeBlocked = {isActingAsDelegate}
    >
        <ScreenWrapper
        .
        .
        </ScreenWrapper>
     </AccessOrNotFoundWrapper>
```
https://github.com/Expensify/App/blob/3946b226d6367a719af5bc894546b95dc9b59fb4/src/pages/settings/Profile/Contacts/NewContactMethodPage.tsx#L103-L152
    

### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by c3024** [comment]
Thanks for the proposal @ChavdaSachin. 

@Expensify/design 

Should we hide the "New Contact Method" button altogether or disable the button with a message on hover when a copilot accesses the account they are delegated to?

**Comment by dubielzyk-expensify** [comment]
Hmm. We tend to not do disabling or hiding in this kind of situation, but I also think the error is enough. Keen to hear if @dannymcclain agrees.

**Comment by dannymcclain** [comment]
I do agree. And if the problem is that the copilot isn't authorized to take that action, we already have a pattern for handling that:
![CleanShot 2024-10-08 at 09 10 35@2x](https://github.com/user-attachments/assets/d0552961-1ad4-44c3-a6e8-38f07b7691a5)

I realize that the above is showing a restricted action because the copilot is set to `Limited`, but I think the same pattern can easily apply here if permissions are the issue. We could just update the copy of the modal to be something like:

>**Not so fast...**
>You don't have permission to take this action for {{user}} as a copilot.

I'm imagining this modal would pop up when the copilot clicks the `New contact method` button. Thoughts?

**Comment by ChavdaSachin** [comment]
[Updated Proposal](https://github.com/Expensify/App/issues/50063#issuecomment-2388882389)
added @dannymcclain 's suggestion as alternate solution.

**Comment by dubielzyk-expensify** [comment]
Love that @dannymcclain !

**Comment by anmurali** [comment]
@c3024 - can you please review @ChavdaSachin revised proposal?

**Comment by c3024** [comment]
@ChavdaSachin’s alternative [solution](https://github.com/Expensify/App/issues/50063#issuecomment-2388882389), as per design inputs, looks good to me. We can work on more specifics in the PR with reviews from the design team.

🎀 👀 🎀 C+ reviewed.

**Comment by ChavdaSachin** [comment]
@Expensify/design
Could I have input on whether we want to change anything about 'Not So Fast....' prompt?
since the prompt has text '.....Limited Access copilot', and here for this problem we are gonna use this same prompt for Full Access Copilot as well. and there is no such different implementation for Full Access Copilot.

**Comment by dannymcclain** [comment]
If it's going to be the exact same implementation across everything I think we should just change it everywhere to something like what I mentioned [here](https://github.com/Expensify/App/issues/50063#issuecomment-2399975948):

>Not so fast...
>You don't have permission to take this action for {{user}} as a copilot.

@dylanexpensify @dubielzyk-expensify (and whoever else I should tag 😅) are y'all good with that? I see in the original we have `limited access` linked... Not sure where that goes, but if we need to maintain the link, would it still make sense to make the word `copilot` the link text?

**Comment by ChavdaSachin** [comment]
This is the redirect link - https://help.expensify.com/expensify-classic/hubs/copilots-and-delegates
Link seems good enough for all kind of copilots. ✅

**Comment by dannymcclain** [comment]
Oh nice. 👍

**Comment by dubielzyk-expensify** [comment]
I'm okay with that change 👍

**Comment by ChavdaSachin** [comment]
Could anyone add my following accounts to copilot beta....
sachin.expensify.test+51@gmail.com
sachin.expensify.test+24@gmail.com
Thank you.

**Comment by ChavdaSachin** [comment]
Hey @Julesssss I found the exact same issue,
When trying to `issue a new Expensify card` from copilot account - I got following response from BE 
<details><summary> response </summary>
{
    "code": 666,
    "jsonCode": 666,
    "type": "Expensify\\Libs\\Error\\ExpError",
    "UUID": "7D862FFF-854D-4AE0-B50B-6C8C96D37AB9",
    "message": "You must be a domain admin to create the virtual card.",
    "title": "",
    "data": [],
    "htmlMessage": "",
    "onyxData": [],
    "requestID": "8d2f62629dff8531-BOM"
}
</details>

Above response clearly suggests that copilot would not be able to issue the card.
So IMO we should implement exact same solution for `issue new Expensify card` flow,
additionally we need to check for company cards as well.
(I do not have test card data so I was unable to verify whether BE returns error for company card or not).

cc. @c3024

**Comment by Julesssss** [comment]
> So IMO we should implement exact same solution for issue new Expensify card flow,
additionally we need to check for company cards as well.

Yeah I agree 👍

**Comment by Julesssss** [comment]
Added to beta. Here's the new issue.

> (I do not have test card data so I was unable to verify whether BE returns error for company card or not).

We want to prevent this for company cards too.

**Comment by Julesssss** [comment]
Hey @c3024 and @ChavdaSachin can you please comment on that issue so I can assign you. Thanks

**Comment by c3024** [comment]
> **BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
> 
> * [x]  [@c3024] The PR that introduced the bug has been identified. Link to the PR: The frontend part was not implemented for this yet. So, no PR can be held responsible for this.
> * [x]  [@c3024] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: NA
> * [x]  [@c3024] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: No discussion was started because this could not have been identified earlier.
> * [x]  [@c3024] Determine if we should create a regression test for this bug. Yes
> * [x]  [@c3024] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

### Regression Test Proposal
1. Navigate to "Settings" > "Profile" as a copilot.
2. Click  on 'Contact methods'.
3. Click on 'New contact method' button.
4. Verify 'Not So Fast...' modal appears and 'New contact method' flow does not start.

**Comment by Julesssss** [comment]
Awaiting payment

**Comment by anmurali** [comment]
@ChavdaSachin - can you please give me your Upwork profile? You are in Slack, yes?
@c3024 is paid.

**Comment by ChavdaSachin** [comment]
I am not on slack yet.
https://www.upwork.com/freelancers/~016a0f0a12bce54a49

I have already applied for the job on upwork.

**Comment by ChavdaSachin** [comment]
@anmurali can you get it done today?
That would be helpful.

**Comment by anmurali** [comment]
Offer is [here](https://www.upwork.com/nx/wm/offer/104760829)

**Comment by ChavdaSachin** [comment]
Accepted

**Comment by ChavdaSachin** [comment]
friendly bump ^ @anmurali
